Mathlib/GroupTheory/Submonoid/Operations.lean

```diff
@@ -1405,6 +1405,18 @@ theorem bot_or_exists_ne_one (S : Submonoid M) : S = ⊥ ∨ ∃ x ∈ S, x ≠
 
 end Submonoid
 
+/-- The `MonoidHom` of submonoids induced by a `MonoidHom` of monoids. -/
+@[to_additive
+      "The `AddMonoidHom` of additive submonoids induced by an `AddMonoidHom` of additive monoids."]
+def MonoidHom.submonoidMap' {M' : Submonoid M} {N' : Submonoid N} (f : M →* N)
+    (hf : M'.map f ≤ N') : M' →* N' :=
+  (Submonoid.inclusion hf).comp <| f.submonoidMap M'
```

Comment:
We have room to do better in terms of naming. I suggest we follow the pattern already in place for linear maps and rename [MonoidHom.restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/Submonoid/Operations.html#MonoidHom.restrict) to `MonoidHom.domRestrict` so that we can use that name here.

Also, it is usually better to make statements using pull backs rather than push forwards so the hypothesis `hf` should be stated as `Set.MapsTo f M' N'` (note that this is defeq to `∀ x in M', f x ∈ N'` as well as `M' ≤ N'.comap f`).

I thus suggest:
```suggestion
/-- Restrict domain and codomain of a morphism of monoids. -/
@[to_additive "Restrict domain and codomain of a morphism of additive monoids."]
def MonoidHom.restrict {M' : Submonoid M} {N' : Submonoid N} (f : M →* N)
    (hf : Set.MapsTo f M' N') : M' →* N' :=
  (f.domRestrict M').codRestrict N' <| SetLike.forall.2 hf
```
Likewise for the `Group` version.